File: crates/mobee-core/src/kinds.rs
//! The single registry of every mobee nostr event kind.
//!
//! This module is the ONE place a kind *number* may appear. Every other site refers to a named
//! constant from here (re-exported through [`crate::gateway`] for the trade-path kinds). The
//! trade kinds form a contiguous mobee-owned block — `3400`–`3405` — plus the addressable
//! seller heartbeat `30340`.
//!
//! | Kind | Object | Author |
//! |---|---|---|
//! | `3400` | RECEIPT | buyer + seller (co-signed) |
//! | `3401` | OFFER | buyer |
//! | `3402` | CLAIM (bid + `creq` invoice) | seller |
//! | `3403` | RESULT | seller |
//! | `3404` | FEEDBACK (progress / error / refusal) | seller |
//! | `3405` | AWARD (claim selection) | buyer |
//! | `30340` | SELLER HEARTBEAT (addressable, `d="mobee-seller"`) | seller |

/// Co-signed settlement receipt (buyer + seller).
pub const JOB_RECEIPT_KIND: u16 = 3400;
/// Buyer-authored work offer.
pub const JOB_OFFER_KIND: u16 = 3401;
/// Seller-authored claim carrying the NUT-18 `creq` invoice.
pub const JOB_CLAIM_KIND: u16 = 3402;
/// Seller-authored typed delivery.
pub const JOB_RESULT_KIND: u16 = 3403;
/// Seller-authored progress / error / refusal feedback.
pub const JOB_FEEDBACK_KIND: u16 = 3404;
/// Buyer-authored claim award / acceptance — e-tags the offer + winning claim.
/// A buyer-authored selection must not ride the seller's feedback kind.
pub const JOB_AWARD_KIND: u16 = 3405;
/// Addressable seller liveness heartbeat, `d="mobee-seller"`. Must stay in the NIP-01
/// parameterized-replaceable range `30000`–`39999`, hence `30340` (not a `34xx` value).
pub const SELLER_HEARTBEAT_KIND: u16 = 30340;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn trade_path_kinds_are_the_contiguous_mobee_block() {
        // The trade path lives in a contiguous mobee-owned block; none reuse the generic DVM range.
        assert_eq!(
            [
                JOB_RECEIPT_KIND,
                JOB_OFFER_KIND,
                JOB_CLAIM_KIND,
                JOB_RESULT_KIND,
                JOB_FEEDBACK_KIND,
                JOB_AWARD_KIND,
            ],
            [3400, 3401, 3402, 3403, 3404, 3405]
        );
        for kind in [5109u16, 6109, 7000] {
            assert!(
                ![
                    JOB_RECEIPT_KIND,
                    JOB_OFFER_KIND,
                    JOB_CLAIM_KIND,
                    JOB_RESULT_KIND,
                    JOB_FEEDBACK_KIND,
                    JOB_AWARD_KIND
                ]
                .contains(&kind),
                "generic DVM kind {kind} must not appear in the mobee block"
            );
        }
    }

    #[test]
    fn heartbeat_is_addressable_replaceable() {
        assert!((30000..=39999).contains(&SELLER_HEARTBEAT_KIND));
    }
}

File: web/network/js/kinds.js
/**
 * SINGLE SOURCE OF TRUTH for every Nostr `kind` number the observatory touches.
 *
 * This file is the ONLY place a kind literal may appear — every other module imports
 * the named constants below and MUST NOT hard-code a kind number. A grep test
 * (test/kinds.test.mjs) enforces that: the marketplace kind digits appear nowhere in
 * js/ or scripts/ except here.
 */

/** NIP-01 profile metadata. Nostr-standard — carries no mobee tag. */
export const PROFILE = 0;

/** Job offer the buyer posts. Sellers claim it. */
export const OFFER = 3401;
/** Seller claim — carries the NUT-18 payment request (`creq`). The seller bids to do the job. */
export const CLAIM = 3402;
/** Seller result — the delivery. */
export const RESULT = 3403;
/** Seller feedback — a progress / error / refusal note on a job. */
export const FEEDBACK = 3404;
/** Buyer award — selects a claim, e-tagging the offer and the winning claim. */
export const AWARD = 3405;
/** Co-signed payment receipt — the settlement proof. */
export const RECEIPT = 3400;
/** NIP-89 seller handler announce (a seller capability advert). Carries no mobee tag. */
export const HANDLER = 31990;
/**
 * Seller liveness heartbeat. Addressable (parameterized-replaceable): keyed by
 * (author, kind, d) — resolve the current one by AUTHOR + KIND (+ d), taking the
 * newest created_at. NEVER look it up by a published event id (a replaceable event
 * is superseded, so by-id lookups go empty and read as a false "offline").
 */
export const HEARTBEAT = 30340;

/** The mobee namespace tag value. Every trade event and the heartbeat carry `["t","mobee"]`. */
export const MOBEE_TAG = "mobee";

/** Plain-English labels for a kind, for any place a kind must surface to a human. */
export const KIND_LABELS = Object.freeze({
  [PROFILE]: "profile",
  [OFFER]: "offer",
  [CLAIM]: "claim",
  [RESULT]: "result",
  [FEEDBACK]: "feedback",
  [AWARD]: "award",
  [RECEIPT]: "receipt",
  [HANDLER]: "handler (NIP-89)",
  [HEARTBEAT]: "heartbeat",
});

/**
 * Marketplace kinds that carry `["t","mobee"]` — requested with a `#t:["mobee"]` filter.
 * The trade path plus the seller heartbeat all live in the mobee namespace.
 */
export const MOBEE_TAGGED_KINDS = Object.freeze([
  OFFER,
  CLAIM,
  RESULT,
  FEEDBACK,
  AWARD,
  RECEIPT,
  HEARTBEAT,
]);

/**
 * Marketplace kinds requested WITHOUT a t-tag filter — the NIP-89 handler announce is a
 * standard advert that carries no mobee tag, so a `#t` filter would hide it. Gift-wrap
 * (1059) stays dark either way.
 */
export const UNTAGGED_KINDS = Object.freeze([HANDLER]);
